### PR TITLE
SecretStore documentation update

### PR DESCRIPTION
**Pull request by svyatonik**
I do not remember - why I haven't opened PR a month ago - looks like everything is OK with it.

In addition to docs update, I've also added an illustration to the glossary - still hope that after reading this section, it should be easier to understand different SS aspects.

### CODE REVIEW

File: Secret-Store-Configuration.md [hunk 1]
```diff
@@ -56,9 +56,18 @@ nodes = ["165872fc7999f12f8d5a4156d8cf95cdd363ab2dffb53effcef926863beead042e926d
...
 
-The main disadvantage of hardcoded list is that if you need to change nodes set, you have to stop all nodes, change their configuration files and finally restart. Another option is to use blockchain contract to configure nodes set (see [Nodes set contracts](#nodes-set-contracts)).
+The main disadvantage of hardcoded list is that if you need to change nodes set, you have to stop all nodes, change their configuration files and finally restart. Another option is to use blockchain contract to configure the set (see [Nodes set contracts](#nodes-set-contracts)), by using following option in the configuration file:
+```toml
+[secretstore]
+server_set_contract = "registry"
+```
 
-### Configuring Secret Store API
+There are four possible values for this option:
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: three?!

Comment (svyatonik) [reply Secret-Store-Configuration.md]: right, thanks :)

File: Secret-Store-Configuration.md [hunk 2]
```diff
@@ -56,9 +56,18 @@ nodes = ["165872fc7999f12f8d5a4156d8cf95cdd363ab2dffb53effcef926863beead042e926d
...
+- `"none"` - nodes set will be read from the `nodes` configuration option;
+- `"registry"` - nodes set contract address will be read from `server_set_contract` entry in the Registry;
+- `address` - this node will be configured for the nodes set contract, residing at given address.
+
+### Configuring Secret Store HTTP API
 Secret Store HTTP API is enabled by default on `local` interface and port `8082`. To change these settings, use following configuration options:
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: I think it is more precise with `localhost` here instead of `local`

File: Secret-Store-Configuration.md [hunk 3]
```diff
@@ -69,7 +78,40 @@ http_port = 12001
 
 You can disable HTTP API by setting `disable_http` option to `true`.
 
-The same API can be provided using blockchain contract (see [Secret Store API contract](#secret-store-api-contract)).
+### Configuring Secret Store service contract API
+As an aternative way to provide Secret Store API, you can deploy a set of service contracts on the blockchain and Secret Store will serve requests, coming to these contracts. There are currently four configuration options for these contracts:
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: s/aternative/alternative

File: Secret-Store-Configuration.md (hunk 2 above)
Comment (svyatonik) [reply Secret-Store-Configuration.md]: That's also the possible value of the param (i.e. you're passing `local` as `http_interface` value -  not the `localhost`). I'd prefer to leave `local` to not to confuse users :) Or maybe reword it somehow? What do you think is better?

Comment (niklasad1) [reply Secret-Store-Configuration.md]: Ok, make sense then it's ok for me :+1:

File: Secret-Store-Configuration.md [hunk 4]
```diff
@@ -69,7 +78,40 @@ http_port = 12001
...
+As an alternative way to provide Secret Store API, you can deploy a set of service contracts on the blockchain and Secret Store will serve requests, coming to these contracts. There are currently four configuration options for these contracts:
+```toml
+[secretstore]
+service_contract_srv_gen = "50D59af572FF790D665C339E60128C511c2656f0"
+service_contract_srv_retr = "registry"
+service_contract_doc_store = "none"
+service_contract_doc_sretr = "registry"
+```
+
+Every service contract is responsible for providing access to the single type of sessions:
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: Hrm, I guess that session implies configuration here?

It is not clear to me before first you introduce `there are four configurations` then `the single type of sessions`.

Also, I prefer to rephrase `to the single type of sessions` to something like `to exactly one of the session types`

File: Secret-Store-Configuration.md [hunk 5]
```diff
@@ -69,7 +78,40 @@ http_port = 12001
...
+
+Every option can have following values:
+- `"none"` - this node won't serve requests coming from this kind of a contract;
+- `"registry"` - contract address will be read from registry (entries are named: `secretstore_service_srv_gen`, `secretstore_service_srv_retr`, `secretstore_service_doc_store` and `secretstore_service_doc_sretr` accordingly);
+- `address` - this node will listen to requests from the contract, residing at given address.
+
+For more information on how every contract works, please refer to [Secret Store service contracts](#secret-store-service-contracts)
+
+### Configuring Secret Store permissioning contract
+There's an option to configure address of permissioning contract in Secret Store:
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: s/permissioning contract/the permissioning contract

File: Secret-Store-Configuration.md (hunk 4 above)
Comment (svyatonik) [reply Secret-Store-Configuration.md]: Rephrased. Not sure what do you mean by "session implies configuration", though :) You're deploying four contracts - every contract has a some method `start`. You call `contract1.start()` - it is starting session of type1. You call `contract2.start()` - it is starting session of type2. Maybe add some hints here? Is this not clear?

File: Secret-Store-Configuration.md [hunk 6]
```diff
@@ -69,7 +78,40 @@ http_port = 12001
...
+```toml
+[secretstore]
+acl_contract = "registry"
+```
+
+This option can have 3 values:
+- `"none"` - node will be configured to work without actual permissions check. This means that it will hand out its keys shares to anyone who will request it. **This mode is intended for tests and must NOT be used in production**;
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: if my grammar skills are correct it should be (because permissions are plural):

s/permissions check/permission checks

Comment (niklasad1) [reply Secret-Store-Configuration.md]: s/keys shares/key shares

File: Secret-Store-Configuration.md [hunk 7]
```diff
@@ -152,56 +203,152 @@ The `disable_auto_migrate` must be set to `true` in configuration file, if you'r
...
+	function isMigrationConfirmed(address keyServer) external view returns (bool);
+	/// Start migration.
+	function startMigration(bytes32 id) external;
+	/// Confirm migration.
+	function confirmMigration(bytes32 id) external;
 }
 ```
 
-For example implementation, see [SetOwnedWithMigration.sol](https://github.com/svyatonik/contracts/blob/94891dbacf2be8edc497396e41add21117b63359/contracts/SetOwnedWithMigration.sol). The `disable_auto_migrate` must be set to `false` in configuration file, if you're using this type of the contract.
+For implementation of this contract, please refer to [SetOwnedWithMigration.sol](https://github.com/parity-contracts/secretstore-key-server-set/blob/cf2d4cb4becf7a71686b2d66fa34ac4c5a46d610/contracts/SetOwnedWithMigration.sol). The `disable_auto_migrate` must be set to `false` in configuration file, if you're using this type of the contract.
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: s/if you're using this type of the contract./if you're using this contract  ?!

File: Secret-Store-Configuration.md [hunk 8]
```diff
@@ -152,56 +203,152 @@ The `disable_auto_migrate` must be set to `true` in configuration file, if you'r
...
+
+Please notice, that nodes are sending transactions to this contract. So make sure that every node (from 'current', 'new' and 'migration' sets) has enough ETH to cover the gas costs. Otherwise, migration can never start or never complete.
+
+## Secret Store service contracts
+There are two kind of participants, sending transactions to service contracts:
+- SS clients, wanting to access SS API;
+- SS nodes, responding to service contracts requests.
+
+This implies different contract interfaces for different participant types - one is 'ClientApi' and the other is 'KeyServerApi'. Key servers are sending transactions to service contract, so please make sure that key servers accounts has enough ether to cover gas costs. [Reference implementation]
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: s/service contract/the service contract

File: Secret-Store-Configuration.md [hunk 9]
```diff
@@ -152,56 +203,152 @@ The `disable_auto_migrate` must be set to `true` in configuration file, if you'r
...
+(https://github.com/parity-contracts/secretstore-service/pull/1/files) of service contracts requires a fee to be paid for every request. The fee is splitted among all key servers and can be transferred to their accounts by calling 'drain' method. This can be used as a source of ETH to pay transactions costs.
+
+Here we will consider 'ClientApi' interfaces only. One common thing about these interfaces is that they all have single method to register request and one or two events, signalling that request has been served. There's always additional event, signalling that request has failed. It doesn't log the real reason behind the failure, though - you can only find it in nodes logs.
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: Consider rephrasing: `There's always additional event` to `There's always an additional event, ....`  or additional events!

File: Secret-Store-Configuration.md [hunk 10]
```diff
@@ -152,56 +203,152 @@ The `disable_auto_migrate` must be set to `true` in configuration file, if you'r
...
+	/// When error occurs during server key generation.
+	event ServerKeyGenerationError(bytes32 indexed serverKeyId);
+
+	/// Request new server key generation. Generated key will be published via ServerKeyGenerated event when available.
+	function generateServerKey(bytes32 serverKeyId, uint8 threshold) external payable;
+}
+```
+
+### Server Key retrieval service contract
+The client API is straightforward - just call the `retrieveServerKey` method and wait until key is either retrieved (`ServerKeyRetrieved`), or generation error occurs (`ServerKeyRetrievalError`):
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: s/or generation error occurs/or an error occurs

File: Secret-Store-Configuration.md [hunk 11]
```diff
@@ -152,56 +203,152 @@ The `disable_auto_migrate` must be set to `true` in configuration file, if you'r
...
+### ECDSA signature generation contract
+*In development*
 
-To try it, please deploy contract from [SecretStoreServiceFixed.sol](https://github.com/svyatonik/contracts/blob/94891dbacf2be8edc497396e41add21117b63359/SecretStoreServiceFixed.sol) and register it under `secretstore_service` name in the Registry. Add following configuration option to all nodes configuration files: `service_contract = "registry"`. Now you can try to call `generateServerKey` method on the contract and public portion of the server key will be published via `ServerKeyGenerated` event within few blocks.
+### Service contracts on Kovan
+All service contracts, described above, are deployed on Kovan testnet (**testnet means that it should be used for tests only - please do not rely on this as an infrastructure for your non-test applications**). This gives you an ability to store encryption keys even without deploying your own Secret Store. Here's the information required to use these contracts:
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: on the Kovan testnet

File: Secret-Store-Configuration.md [hunk 12]
```diff
@@ -152,56 +203,152 @@ The `disable_auto_migrate` must be set to `true` in configuration file, if you'r
...
+- processing requests could take some time (up to 10 blocks for [Document Key shadow retrieval](#document-key-shadow-retrieval-service-contract) requests and 4-5 blocks for all other requests). It could be even more, when key server misses the request (because of synchronization). The general rule - if your request haven't processed for >100 blocks, then something is definitely wrong;
+- you are not the only user of this Secret Store, so expect keys with IDs like '0x0000000000000000000000000000000000000000000000000000000000000001' to be occupied and some requests to fail because of this.
 
-Please note that for this contract to work, nodes keys must be the same in configuration file and in `SecretStoreService` constructor.
+If you experiencing any problems with service contracts on Kovan, please report at [Parity Support](https://riot.im/app/#/room/#support:matrix.parity.io).
```
Comment (niklasad1) [inline Secret-Store-Configuration.md]: If you experience

File: Secret-Store.md [hunk 13]
```diff
@@ -33,6 +33,32 @@ Terms that are being used on this page:
...
+
+There are 4 entities, which are not a part of **Secret Store**:
+- `Alice` and `Bob` are clients of **Secret Store**, which are using its API to read and store keys;
+- **permissioning contract**, which contains rules that:
+	- allow `Alice` to run **permissioning sessions** with server key with **id** = 1. Since **server key** 1 doesn't have **document key** bound to it, it only makes sense for Alice to run **signing sessions** to sign messages with the private portion of key 1 (`PRIV1`);
+	- allow `Bob` to run **permissioning sessions** with server key with **id** = 2. Since **server key** 1 has **document key** bound to it, `Bob` can run **document key retrieval sessions** (to use `PRIV2` to decrypt `D2` and return it, encrypted with `Bob`' private key) along with **signing sessions** (to sign messages using `PRIV2`);
+- **node** (`NODE4`), that is not currently a part of **Secret Store**.
+
+Following are examples of API calls that will succeed:
+- `Alice` runs [**server key generation session**](#server-key-generation-session) to generate new **server key** with **id** = 3;
```
Comment (niklasad1) [inline Secret-Store.md]: a new **server key** / a new **server key pair**

File: Secret-Store-Configuration.md (hunk 6 above)
Comment (svyatonik) [reply Secret-Store-Configuration.md]: 'key shares' means that there's one key - that's wrong

File: Secret-Store-Configuration.md (hunk 7 above)
Comment (svyatonik) [reply Secret-Store-Configuration.md]: Not sure what's better - we're talking here about **any** contract that is providing given API

File: Secret-Store.md (hunk 13 above)
Comment (svyatonik) [reply Secret-Store.md]: It is called **server key** across the docs && specifically in the glossary.

**Comment by Tbaut** [comment]
Arg, I was working on it, will submit a new PR.
